File: PDOO/p4/Civitas/src/civitas/JugadorEspeculador.java

package civitas;
public class JugadorEspeculador extends Jugador {
    private static int FactorEspeculador = 2;
    private float fianza;
    private Boolean especulador;
    
    JugadorEspeculador(Jugador jugador, float fianza){
        super(jugador);
        this.fianza = fianza;
        especulador = true;
        for (int i = 0; i < super.getPropiedades().size(); i++){
            super.getPropiedades().get(i).actualizaPropietarioPorConversion(this);
        }
    }
    
    @Override
    Boolean encarcelar(int numCasillaCarcel){
        Boolean resultado = false;
        if (!super.encarcelado){
            if (!super.tieneSalvoconducto()){
                if (fianza >= super.getSaldo()){
                    resultado = true;
                } else {
                    super.paga(fianza);
                    Diario.getInstance().ocurreEvento("El jugador "+super.getNombre()+" Se ha librado de la carcel paganzo una fianza");
                }
            } else {
                resultado = false;
                super.perderSalvoconducto();
                Diario.getInstance().ocurreEvento("El jugador "+super.getNombre()+" se libra de la carcel");
            }
        }
        return resultado;
    }
    
    @Override
    int getHotelesMax(){
        return (super.getHotelesMax() * FactorEspeculador );
    }
    
    @Override
    int getCasasMax(){
        return (super.getCasasMax() * FactorEspeculador );
    }
    
    @Override
    Boolean pagaImpuesto(float cantidad){
        Boolean resultado = false;
        if (!encarcelado){
            float impuesto = cantidad / 2;
            resultado = paga(impuesto);
            Diario.getInstance().ocurreEvento("El jugador "+super.getNombre()+" paga "+impuesto+" euros por impuesto");
        }
        return resultado;
    }
    
    @Override
    public String toString(){
        return "Jugador{ nombre=" + super.getNombre() + ", encarcelado=" + Boolean.toString(super.isEncarcelado()) +  ", numCasillaActual=" + super.getNumCasillaActual() + ", puedeComprar=" + super.getPuedeComprar() +
                ", saldo=" + super.getSaldo() + ", especulador=" + Boolean.toString(especulador) + "}";
    }
}

File: PDOO/p4/civitas_rb/lib/jugador_especulador.rb
module Civitas

class Jugador_especulador < Jugador
  @@factorespeculador=2
  
  def initialize(jugador, fianza)
    super(jugador)
    @fianza=fianza
    @especulador=true
    for i in 0...super.propiedades.length
      super.propiedades.at(i).actualizar_propietario_por_conversion(self)
    end
  end
  
  def encarcelar(numCasillaCarcel)
    resultado = false
    if !super.encarcelado
      if !super.tiene_salvoconducto
        if @fianza >= super.saldo
          resultado=true
        else
          super.paga(@fianza)
          Diario.instance.ocurre_evento("El jugador #{super.nombre} se ha librado de la carcel pagando una fianza")
        end
      else
        resultado=false
        super.perder_salvoconducto
        Diario.instance.ocurre_evento("El jugador #{super.nombre} se libra de la carcel")
      end
    end
    return resultado
  end
  
  def gethotelesmax
    return (super.hotelesMax * @@factorespeculador)
  end
  
  def getcasasmax
    return (super.casasMax * @@factorespeculador)
  end
  
  def paga_impuesto(cantidad)
    resultado=false
    if !super.encarcelado
      impuesto = cantidad / 2
      Diario.instance.ocurre_evento("El jugador #{super.nombre} paga #{impuesto} euros por impuesto")
    end
    return resultado
  end
  
  def to_string
    return "Jugador{ nombre= #{super.nombre}, encarcelado= #{super.is_encarcelado.to_s}, numCasillaActual= #{super.numCasillaActual}, puedeComprar= #{super.puedeComprar}, 
            saldo= #{super.saldo}, especulador= #{@especulador.to_s} }"
  end
  
  
end
end